pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -312,8 +312,58 @@ public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentit
+    // Get the per-query timeout from the broker request query options if exists
+    long queryTimeoutMs = QueryOptions.getTimeoutMs(brokerRequest.getQueryOptions());
+    // Remaining time in milliseconds for the server query execution
+    // Use the max of offline table remaining time and realtime table remaining time for hybrid use case
+    long remainingTimeMs = 0;
+    // NOTE: Use query-level timeout if exists, or use table-level timeout if exists, or use instance-level timeout
+    if (offlineBrokerRequest != null) {
+      long offlineTimeoutMs;
+      if (queryTimeoutMs > 0) {
+        offlineTimeoutMs = queryTimeoutMs;
+      } else {
+        Long tableTimeoutMs = _routingManager.getTimeoutMs(offlineTableName);
+        offlineTimeoutMs = tableTimeoutMs != null ? tableTimeoutMs : _brokerTimeoutMs;
+      }
+      long offlineRemainingTimeMs = offlineTimeoutMs - timeSpentMs;
+      if (offlineRemainingTimeMs <= 0) {
+        LOGGER.info("Offline table timed out (time spent: {}ms, timeout: {}) before scattering the request {}: {}",
+            timeSpentMs, offlineTimeoutMs, requestId, query);
+        _brokerMetrics.addMeteredTableValue(rawTableName, BrokerMeter.REQUEST_TIMEOUT_BEFORE_SCATTERED_EXCEPTIONS, 1);
+        return new BrokerResponseNative(QueryException.getException(QueryException.BROKER_TIMEOUT_ERROR, String
+            .format("Offline table timed out (time spent: %dms, timeout: %dms) before scattering the request",
+                timeSpentMs, offlineTimeoutMs)));
+      }
+      offlineBrokerRequest.getQueryOptions()
+          .put(Broker.Request.QueryOptionKey.TIMEOUT_MS, Long.toString(offlineRemainingTimeMs));
+      remainingTimeMs = offlineRemainingTimeMs;
+    }
+    if (realtimeBrokerRequest != null) {
+      long realtimeTimeoutMs;
+      if (queryTimeoutMs > 0) {
+        realtimeTimeoutMs = queryTimeoutMs;
+      } else {
+        Long tableTimeoutMs = _routingManager.getTimeoutMs(realtimeTableName);
+        realtimeTimeoutMs = tableTimeoutMs != null ? tableTimeoutMs : _brokerTimeoutMs;
+      }
+      long realtimeRemainingTimeMs = realtimeTimeoutMs - timeSpentMs;
+      if (realtimeRemainingTimeMs <= 0) {
+        LOGGER.info("Realtime table timed out (time spent: {}ms, timeout: {}) before scattering the request {}: {}",
+            timeSpentMs, realtimeTimeoutMs, requestId, query);
+        _brokerMetrics.addMeteredTableValue(rawTableName, BrokerMeter.REQUEST_TIMEOUT_BEFORE_SCATTERED_EXCEPTIONS, 1);
+        return new BrokerResponseNative(QueryException.getException(QueryException.BROKER_TIMEOUT_ERROR, String
+            .format("Realtime table timed out (time spent: %dms, timeout: %dms) before scattering the request",
+                timeSpentMs, realtimeTimeoutMs)));
+      }
+      realtimeBrokerRequest.getQueryOptions()
+          .put(Broker.Request.QueryOptionKey.TIMEOUT_MS, Long.toString(realtimeRemainingTimeMs));
+      remainingTimeMs = Math.max(remainingTimeMs, realtimeRemainingTimeMs);
```

Comment:
Math.min? Should we not take the stricter time to be the timeout value?

Reply:
Should be max. Updated the comments as following:
```
    // NOTE: Use the max of offline table remaining time and realtime table remaining time for hybrid use case if the
    //       timeout for them are not configured to be the same. Server side will have different remaining time set for
    //       each table type, and broker should wait for both types to return.
```